### DISCUSSION TITLE
Retrieving `Secret`s in `5.0.0`

### DISCUSSION THREAD

**Discussion by mehroozkhan** [sbertix/Swiftagram#182] [Q&A] [answered]
Hi , 

I update the library, but now I can't access secret, can you please help, 

I use to get secret like this : 

`let secret = KeychainStorage<Secret>().all().first`

this is to check if user is already login. 

but with the update its giving me an error: can't find keychainstorage in scope

**Comment by sbertix** [comment, marked as answer]
Hey @mehroozkhan,

You can either import **ComposableStorageCrypto** or use the new authentication flow. 

```swift
let secret = try? Authenticator.keychain.secrets.get().first
``` 

Check out the _Followers_ example to see more use cases.

**Reply by mehroozkhan** [reply]
Hi @sbertix : 

when I try to get secret , it throws an exception : 
```
DecodingError
  ▿ typeMismatch : 2 elements
    - .0 : Swift.Dictionary<Swift.String, Any>
    ▿ .1 : Context
      - codingPath : 0 elements
      - debugDescription : "Expected to decode Dictionary<String, Any> but found a number instead."
      - underlyingError : nil
```

**Reply by sbertix** [reply]
Do you mind sharing the line causing this error? 

Tbh it doesn't look like it's something related to `Secret`... there are no numbers and no dictionaries in it... 🤔

What is it that you're exactly doing in your code?

**Reply by mehroozkhan** [reply]
my code :  

`do {
            let secret = try Authenticator.keychain.secrets.get().first
        }
        catch {
            print(error)
        }`
        
also `KeychainStorage<Secret>().all().first` gives me same error

**Reply by mehroozkhan** [reply]
code I am using to login is: 

```
Authenticator.keychain
            .visual {
                self.webView = $0
            }
            .authenticate()
            .sink { [weak self] in
                switch $0 {
                case .failure(let error):
                    print(error.localizedDescription)
                    self?.loginBlock?(false)
                default:
                    self?.loginBlock?(true)
                    break
                }
            } receiveValue: { secret in
                
            }.store(in: &self.bin)
```

**Reply by mehroozkhan** [reply]
and now its not even letting me login, giving me an error : 

`The specified attribute does not exist`

`status(-25303)`

**Reply by sbertix** [reply]
Did you try removing **Swiftagram** completely from your SPM dependencies and adding it back?

**Reply by mehroozkhan** [reply]
just did, still not working

**Reply by sbertix** [reply]
Are you perhaps using your keychain for something else? 
It shouldn't be an issue but I'm trying to figure out how it gets the wrong data to parse.

**Reply by mehroozkhan** [reply]
No, I am not using keychain for anything in my project.

**Reply by mehroozkhan** [reply]
btw it gives me an error right after login as well : 

```
Authenticator.keychain
            .visual {
                self.webView = $0
            }
            .authenticate()
            .sink { _ in
                
            } receiveValue: { secret in
               
                do { let secret = try Authenticator.keychain.secrets.get().first }
                catch {
                    print(error)
                }
            }.store(in: &self.bin)
```

I tried simulator reset as well, still same issue.

**Reply by sbertix** [reply]
Does logging in with `Authenticator.userDefaults` throw the same exception? 
I'm just trying to figure out if it's something related to the keychain or the actual `Secret` definition.

**Reply by mehroozkhan** [reply]
No, After login with `Authenticator.userDefaults` when I try to get secret  `Authenticator.userDefaults.secrets.get()` , no exception but its empty

**Reply by sbertix** [reply]
Hey @mehroozkhan, this should now be fixed. Make sure your update your SPM dependencies and `ComposableRequest` version matches `5.0.8` 💪

**Reply by biox86** [reply]
Hello guys, something strange related to the login process happened to me. Thought about writing about it in here cause the subject is pretty much the same. 

The problem happened just on the simulator trying to login with Authenticator.keychain ( logging in with Authenticator.userDefaults worked perfectly ).

I was constantly getting failure from the authenticator with Swiftchain.Keychain.Error.status(-25303).

After resetting the device the problem disappeared completely and started working correctly.

The snippet of the authenticator is pretty much a basic one:
DispatchQueue.main.asyncAfter(deadline: .now()) {
                Authenticator.keychain
                    .visual(filling: self.view)
                    .authenticate()
                    .receive(on: RunLoop.main)
                    .sink(receiveCompletion: { print($0); self.dismiss(animated: true, completion: nil) },
                          receiveValue: completion)
                    .store(in: &self.bin)
            }

What do you think could have caused this problem?
Could it be related to something saved in the keychain with the old KeychainStorage class from ver 4.x that could have caused some conflicts?
Or maybe could just be the keychain that is not working well on the simulator? 
I know it doesn't really feel like a real bug, but maybe I'm just missing something about keychain logic.

EDIT:
My analysis was wrong. The problem persists and it happens on bot sim and device.

The first login works correctly and it saves the secret inside the keychain, but if I try to login again and re-save it in the keychain the authenticator always responds with that error.

**Reply by sbertix** [reply]
Hey @biox86,

It seems like it might be the same exception that's already been patched. 
Make sure you update your SPM dependencies and `ComposableRequest` is on `5.0.9`. If it still persists afterwards, please open an issue with some more info 💪
